designer/server/src/main/scala/pl/touk/nussknacker/ui/db/timeseries/questdb/QuestDbFEStatisticsRepository.scala

```diff
@@ -0,0 +1,274 @@
+import java.nio.file.StandardOpenOption
+import java.time.Clock
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicReference}
+import java.util.concurrent.{ArrayBlockingQueue, ThreadPoolExecutor, TimeUnit}
+import scala.concurrent.duration.{Duration, FiniteDuration}
+import scala.concurrent.{ExecutionContext, ExecutionContextExecutorService, Future}
+import scala.util.{Failure, Success, Try}
+
+// TODO list:
+// WARNING - for now is not thread safe - multiple request at the same time will kill application
+// 2. Compacting?
+// 4. Configurable directory of db (for now it's tmp directory) and limiting db file space on disk.
+// 5. Collecting statistics should be deactivable by configuration. (If db creation fails provide NoOp DB as fallback)
+// 6. API should have better types (missing domain layer for FE statistics names).
+// 6. Handling errors while creating the QuestDb.
+// 7. Changing table definition and recreate
+// 9. multi tenant metrics -> add todo that if there are many instances we should change diagrams on grafana
+// 10. flush data in close?
+// todo polish this class in next commits
+private class QuestDbFEStatisticsRepository(private val engine: AtomicReference[CairoEngine], private val clock: Clock)(
+    private implicit val ec: ExecutionContextExecutorService
+) extends FEStatisticsRepository[Future]
+    with LazyLogging {
+
+  private val sqlContext = new ThreadAwareObjectPool(() =>
+    new SqlExecutionContextImpl(engine.get(), 1).`with`(AllowAllSecurityContext.INSTANCE, null)
+  )
+
+  private val recordCursorPool = new ThreadAwareObjectPool(() => engine.get().select(selectQuery, sqlContext.get()))
+
+  private val walTableWriterPool = new ThreadAwareObjectPool(() =>
+    engine.get().getWalWriter(engine.get().getTableTokenIfExists(tableName))
+  )
+
+  private val flushDataTask      = new AtomicReference(createFlushDataTask())
+  private lazy val retentionTask = new AtomicReference(createRetentionTask())
+  private val shouldCleanUpData  = new AtomicBoolean(false)
+
+  override def write(statistics: Map[String, Long]): Future[Unit] = withExceptionHandling(() => {
+    val statsWalTableWriter = walTableWriterPool.get()
+    statistics.foreach { entry =>
+      val row = statsWalTableWriter.newRow(currentTimeMicros())
+      row.putStr(0, entry._1)
+      row.putLong(1, entry._2)
+      row.append()
+    }
+    statsWalTableWriter.commit()
+  })
+
+  override def read(): Future[Map[String, Long]] = withExceptionHandling(() => {
+    recordCursorPool
+      .get()
+      .fetch(sqlContext.get()) { record =>
+        val name  = record.getStrA(0).toString
+        val count = record.getLong(1)
+        name -> count
+      }
+      .toMap
+  })
+
+  private def currentTimeMicros(): Long =
+    Math.multiplyExact(clock.instant().toEpochMilli, 1000L)
+
+  private def close(): Unit = {
+    recordCursorPool.clear()
+    walTableWriterPool.clear()
+    sqlContext.clear()
+    flushDataTask.get().close()
```

Comment:
what about the situation that close will be invoked in the middle of initialization and some references won't be initialized yet?

Reply:
Done, I moved synchronized from engine recreation method to recover. Despite the fact the engine recreation was synchronized, the close method should be safe even on no initialized references.